### PR TITLE
add -o StrictHostKeyChecking no

### PR DESCRIPTION
**Pull request by MalloZup**
since in previous test sshd connect already to ssh, we use this flags of ssh  to avoid typing yes, and make the test more stable.

replace wait_idle with,  wait_still_screen();  because wait_idle is no safer for the test.

since in the actually status of test, we type 2 times passwd: one time "yes", and 2nd the real pwd.

In future, it will be more sure to test with a rsa key. this taks was already tracked on tickets opensuse. 

Commit tested with SLE, reviewed on Opensuse (not tested, but same behaviour)

### CODE REVIEW

**Comment by okurz** [comment]
@MalloZup please review the contribution notes for os-autoinst as they also should be applied here, especially the git commit rules, see https://github.com/os-autoinst/os-autoinst#how-to-contribute

**Comment by MalloZup** [comment]
@okurz, what is not ok in my commit?

**Comment by MalloZup** [comment]
A "still screen" could be detected before the system would ask for a password. 
When then? a still screen is when user type the  pwd, before there is no possibility for that.

**Comment by aaannz** [comment]
There is possibility that `wait_still_screen` can return before actual password prompt. `script_run` does not wait for script to finish, just type it in and wait_idle.
Suppose when network is *really* slow. After typing ssh command machine is basically idle. `script_run` finishes. If it will take more then 7s (default time requirement for still screen) to establish connection then `wait_still_screen` finished but password prompt is not there yet.
But considering we are connecting to localhost, I guess probability for this race is low. Anyway, I think check for actual prompt as @okurz suggest is better.

**Comment by MalloZup** [comment]
aaanz, 7s yes for this test are enough.  My fix was a quick fix, to stop failure of the tests-suite we have now. In future, rsa key will modify the test, but at moment i'm making other test, and i cannot focus 100 on this. but i tested it and eventuality that network is slow, on local system, that could be also a product bug.

**Comment by sysrich** [comment]
I don't like this proposed fix because it changes the behaviour of ssh in a way that is unlikely to be done by a user

I would much rather have the problem resolved by adding a check_screen and type_string 'yes' if ssh is being asked to verify the fingerprint

We shouldn't just turn off all host key checking because it's inconvenient in our test - that's not how users are expected to use the product

**Comment by dzedro** [comment]
btw I just found https://github.com/os-autoinst/os-autoinst/blob/master/consoles/console.pm#L54

**Comment by sysrich** [comment]
That makes sense in a helper function (I don't like it..but it makes sense and I cant think of a better way of doing it in a helper function)

It doesn't make sense in a test which is meant to be validating the behaviour of SSH as part of a test

**Comment by MalloZup** [comment]
we have already tested ssh on previous test, so far. test sshd. so i was thinking that,  in this case we would test ssh in xterm behaviour, the mainly function of test is the xterm part, not how the way the user authenticate via ssh

**Comment by okurz** [comment]
still:
 * wait_still_screen is wrong because you don't know what caused it
 * git commit message should be improved to tell what it fixes. you can reference the progress issue

**Comment by coolo** [comment]
the sshCommand in console.pm is to be run in a local xterm connecting to a remote SUT. *and* consoles don't necessarly have access to testapi calls.

**Comment by MalloZup** [comment]
ok, i will make with wait_serial, and change the commit logs.  
how about adding a RSA-key, on console-setup? this make users too :ice_cream:

**Comment by sysrich** [comment]
+1 to wait_serial
+0.5 to the RSA-key - I like the idea of checking the ssh is behaving properly with host key verification at least once

**Comment by coolo** [comment]
me too - we really have to be careful with system modifications.

**Comment by MalloZup** [comment]
So here the commit with the new fix.
I was thinking, to change the whole test, and make it more simply.
 because there are a lot of sleep, and so on.

2 questions: 
1)why if we use xterm see start_programm("xterm"), we are saying to ssh to open an other xterm console? 

ssh  -o  "StrictHostKeyChecking no" -XC root@localhost   

2) If we change like above, we can execute commands via ssh to test xterm, and delete all this sleep, and send keys. If you want to preserve the "magic" needle, i could make an echo with same string, to match the needle. 
 But since we run a xterm on  xterm terminal, i can't figure out the logic of this, and make the to pass commands via ssh very complicated.

File: tests/x11/sshxterm.pm [hunk 1]
```diff
@@ -15,16 +15,20 @@ sub run() {
...
     mouse_hide(1);
     x11_start_program("xterm");
-    script_run("ssh -XC root\@localhost xterm");
-    type_string "yes\n";
-    wait_idle 6;
+
+    # check last incoming ssh traffic, wait until the connection is opened from localhost.
+    # "ESTAB", AND ::1:ssh, AND dport(destination-port) for control with exactitude, that a inc. port from ssh only from localhost is open.
+    # at this moment type password with serenity.
+    script_run("for i in {1..100}; do /usr/sbin/ss -at dport = :ssh | tail -n1 >/dev/$serialdev; sleep 2; done & ");
```
Comment (okurz) [inline tests/x11/sshxterm.pm]: that looks like a very clever way to do … what exactly? I am sorry but I don't get what this is supposed to do.

File: tests/x11/sshxterm.pm [hunk 2]
```diff
@@ -15,16 +15,20 @@ sub run() {
...
+    script_run("ssh  -o  \"StrictHostKeyChecking no\" -XC root\@localhost xterm ");
+    wait_serial('ESTAB.*::1:ssh');
     type_string "$password\n";
     sleep 2;
     for (1 .. 13) { send_key "ret" }
     $self->set_standard_prompt();
     type_string "echo If you can see this text, ssh-X-forwarding  is working.\n";
     sleep 2;
-    assert_screen 'test-sshxterm-1', 3;
+    assert_screen 'test-sshxterm-1', 5;
```
Comment (okurz) [inline tests/x11/sshxterm.pm]: better change to use default timeout, see explanation here: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/945

**Comment by okurz** [comment]
As you asked about the commit message. The detailed description is ok. The subject line could use less dashes and a capital character to start with :-) How about "Make ssh X-forward password typing more robust". Also don't hesitate the mention the progress issue this fixes, e.g. "Fixes http://progress.opensuse.org/...<url>"

About the purpose of the test: I agree the main point is to test the X forwarding, not authentication. Also, I see the problems with the previous approach as it could not be ensured the text is actually typed in the "remote" ssh session not the local console. The wait_serial approach for password entry is very good, about the other stuff with "ss" I'm not sure. Let others give their opinion.

File: tests/x11/sshxterm.pm (hunk 1 above)
Comment (MalloZup) [inline tests/x11/sshxterm.pm]: try the command with your own pc.
1 time with not your localhost, 2 nd time with your own localhost.
my fix know with precision that when you see the ssh message "type password", the tcp on port 22 get ETABLISHED, so you know that somebody want to connect to ssh, and .*::1 is to ensure that this somebody come from your localhost. i think more stable then this is not possible.

Comment (MalloZup) [inline tests/x11/sshxterm.pm]: try the command with your own pc.
1 time with not your localhost, 2 nd time with your own localhost.
my fix know with precision that when you see the ssh message "type password", the tcp on port 22 get ETABLISHED, so you know that somebody want to connect to ssh, and .*::1 is to ensure that this somebody come from your localhost. i think more stable then this is not possible.

File: tests/x11/sshxterm.pm (hunk 2 above)
Comment (MalloZup) [inline tests/x11/sshxterm.pm]: yop, you are right. what is default timeout?

**Comment by MalloZup** [comment]
you mean in body git, or title, to mention the progress issue?

File: tests/x11/sshxterm.pm (hunk 2 above)
Comment (okurz) [inline tests/x11/sshxterm.pm]: for future reference: https://github.com/os-autoinst/openQA/blob/master/docs/WritingTests.asciidoc#api

**Comment by okurz** [comment]
I will give you a full example of how the git commit message could look like:
```
Make ssh X-forward password typing robust

Make sure there is a process establishing a loopback ssh connection
before typing the password avoiding wait_idle or any race condition.

Also add flag "StrictHostKeyChecking" to avoid the accept dialog
for the key which would only appear if the key was not accepted
in before by other tests, e.g. the console tests testing sshd when
available.

Fixes https://progress.opensuse.org/issues/5346
```

As you are at it: Can you replace the sleeps? I consider a sleep
also broken when it is assumed some condition has changed
afterwards. Either it is not necessary at all or we should rather
use one of "wait_serial", "wait_screen_change", "assert_screen", "check_screen".

**Comment by MalloZup** [comment]
:+1:  ok , thx for gitten me (i try a neologism with git) . i had the same thought with sleep, i will see for that, that's why i was asking for change the tests. wait for the others.

**Comment by MalloZup** [comment]
i have also fast the fix for sshd done. should make 2 pull-request?

**Comment by coolo** [comment]
you're no longer testing X applications going through ssh

File: tests/x11/sshxterm.pm [hunk 3]
```diff
@@ -15,20 +21,16 @@ sub run() {
     my $self = shift;
     mouse_hide(1);
     x11_start_program("xterm");
-    script_run("ssh -XC root\@localhost xterm");
-    type_string "yes\n";
-    wait_idle 6;
+    # (1) ss for wait until the socket is opened from localhost.
+    # "ESTAB", AND ::1:ssh, AND dport(destination-port) for control with exactitude, that a inc. port from ssh only from localhost is open.
+    script_run("for i in {1..100}; do /usr/sbin/ss -at dport = :ssh | tail -n1 >/dev/$serialdev; sleep 3; done & ");
```
Comment (coolo) [inline tests/x11/sshxterm.pm]: Can you please write your tests as humans would test? If you want headless unit tests, put them in scripts in /data. 

But this is explicitly a test checking xterm through ssh, *not* a headless test.

**Comment by MalloZup** [comment]
since okurz, wrote a fix more conformed to openqa test, we can close the pr

**Comment by coolo** [comment]
then let's do that - this PR went astray IMO

**Pull request closed without merging**